### PR TITLE
[WEBSITE-1925][WEBSITE-1981][WEBSITE-2002] Refactor people/show timeline

### PR DESCRIPTION
**Pull request by jennaramdenee**
* Added HistoryHelper, to sort roles by date
* Added GroupingHelper, to group incumbencies by role
* Added tests for HistoryHelper and GroupingHelper
* Refactored people/show timeline view
* Amended PeopleController#show to use HistoryHelper and GroupingHelper
* Amended people/show tests
* Implemented new timeline styling

### CODE REVIEW

File: app/controllers/concerns/grouping_helper.rb [hunk 1]
```diff
@@ -0,0 +1,107 @@
+module GroupingHelper
+
+  class GroupedIncumbency
+    attr_accessor :start_date, :end_date, :nodes, :grouped_by, :type
+  end
+
+  # Returns array of grouped, ungrouped and unknown objects
+  def self.group(data, category, property)
+    category = category
```
Comment (mattrayner) [inline app/controllers/concerns/grouping_helper.rb]: ?

File: app/controllers/concerns/grouping_helper.rb [hunk 2]
```diff
@@ -0,0 +1,107 @@
+module GroupingHelper
+
+  class GroupedIncumbency
+    attr_accessor :start_date, :end_date, :nodes, :grouped_by, :type
+  end
+
+  # Returns array of grouped, ungrouped and unknown objects
+  def self.group(data, category, property)
```
Comment (mattrayner) [inline app/controllers/concerns/grouping_helper.rb]: Please add some whitespace

File: app/controllers/concerns/grouping_helper.rb [hunk 3]
```diff
@@ -0,0 +1,107 @@
+module GroupingHelper
```
Comment (mattrayner) [inline app/controllers/concerns/grouping_helper.rb]: Please add YARD documentation throughout

File: app/controllers/concerns/grouping_helper.rb [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    ungrouped = separate_ungrouped_data(grouped_hash)
+    grouped = create_grouped_data_object(grouped_hash)
+    sorted_groups = sort_grouped_nodes_by_date(grouped)
+    aggregate_sorted_data(sorted_groups, ungrouped, unknown)
+  end
+
+  # Returns hash with grouped keys
+  # e.g. { 12345678 => [...], 23456789 => [...], "UNKNOWN" => [...] }
+  def self.grouping_hash(data, category, property)
+    grouped_data = data.group_by do |data_point|
```
Comment (mattrayner) [inline app/controllers/concerns/grouping_helper.rb]: this assignment is not necessary - `data.group_by` does exactly the same thing

File: app/controllers/concerns/grouping_helper.rb [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      group_value
+    end
+  end
+
+  # Create array with all UNKNOWN nodes
+  def self.separate_unknown_data(data_hash)
+    unknown ||= []
+    data_hash.each do |key, value|
+      if key == "UNKNOWN"
+        unknown = value
```
Comment (mattrayner) [inline app/controllers/concerns/grouping_helper.rb]: Should you not be adding to the array? i.e. `unknown << value`

File: app/controllers/concerns/grouping_helper.rb [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+      group_value = data_point.try(category).try(property)
+      group_value = 'UNKNOWN' if group_value.nil?
+      group_value
+    end
+  end
+
+  # Create array with all UNKNOWN nodes
+  def self.separate_unknown_data(data_hash)
+    unknown ||= []
+    data_hash.each do |key, value|
```
Comment (mattrayner) [inline app/controllers/concerns/grouping_helper.rb]: could you not do a simple lookup here? `data_hash['UNKNOWN'].each do |value|`

Comment (mattrayner) [reply app/controllers/concerns/grouping_helper.rb]: Think of the iterations over the data you are doing

File: app/controllers/concerns/grouping_helper.rb [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    unknown, ungrouped = separate_unknown_and_ungrouped_data(grouped_data)
+    grouped = create_grouped_data_object(grouped_data)
+    aggregate_sorted_data(grouped, ungrouped, unknown)
+  end
+
+  # Returns hash with grouped keys
+  # e.g. { 12345678 => [...], 23456789 => [...], "UNKNOWN" => [...] }
+  def self.group_data(data, category, property)
+    data.group_by do |data_point|
+      group_value = data_point.try(category).try(property)
```
Comment (mattrayner) [inline app/controllers/concerns/grouping_helper.rb]: could this be

```ruby
group_value = data_point.try(category).try(property) || 'UNKNOWN'
```

Comment (mattrayner) [reply app/controllers/concerns/grouping_helper.rb]: if so, this method could become

```ruby
def self.group_data(data, category, property)
    data.group_by { |data_point| data_point.try(category).try(property) || 'UNKNOWN' }
end
```

File: app/controllers/concerns/grouping_helper.rb [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    sort_grouped_nodes_by_date(grouped_object, current_node)
+    grouped << grouped_object
+  end
+
+  # Create array with all UNKNOWN nodes and create array with all ungrouped data
+  def self.create_sorted_array(data_hash)
+    sorted_array = []
+    data_hash.each do |key, value|
+      if key == 'UNKNOWN' || value.length == 1
+        sorted_array << separate_unknown_and_ungrouped_data(value)
```
Comment (jennaramdenee) [inline app/controllers/concerns/grouping_helper.rb]: `sorted_array <<  [value]`

File: app/controllers/concerns/grouping_helper.rb [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    grouped_object.nodes.each do |node|
+      # Find node that represents current role
+      current_node = node if node.try(:end_date).nil? || node.end_date.nil?
+      # Find earliest start date and latest end date
+      start_date = node.start_date if start_date.nil? || node.start_date < start_date
+      end_date = node.end_date if !current_node && (end_date.nil? || node.end_date > end_date)
+    end
+
+    # Set properties of the object
+    grouped_object.type = grouped_object.nodes.first.type
```
Comment (jennaramdenee) [inline app/controllers/concerns/grouping_helper.rb]: maybe add a todo comment to check if this is still okay when we have roles and memberships of a committee

File: app/controllers/concerns/grouping_helper.rb [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    grouped_object.grouped_by = key
+    grouped_object.nodes = data_hash[key]
+
+    start_date = nil
+    end_date = nil
+    current_node = nil
+
+    grouped_object.nodes.each do |node|
+      # Find node that represents current role
+      current_node = node if node.try(:end_date).nil? || node.end_date.nil?
```
Comment (jennaramdenee) [inline app/controllers/concerns/grouping_helper.rb]: maybe add a todo RE: memberships and roles within a committee?

File: app/controllers/concerns/grouping_helper.rb (hunk 3 above)
Comment (jennaramdenee) [reply app/controllers/concerns/grouping_helper.rb]: looks like 3N? now (key length serves as an unknown variable)

File: app/controllers/concerns/grouping_helper.rb [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+  # e.g. { 12345678 => [...], 23456789 => [...], "UNKNOWN" => [...] }
+  def self.group_data(data, category, property)
+    data.group_by { |data_point| data_point.try(category).try(property) || 'UNKNOWN' }
+  end
+
+  # Create array with all UNKNOWN nodes and create array with all ungrouped data
+  def self.create_sorted_array(data_hash)
+    sorted_array = []
+    data_hash.each do |key, value|
+      if key == 'UNKNOWN' || value.length == 1
```
Comment (mattrayner) [inline app/controllers/concerns/grouping_helper.rb]: I'd add a comment here making it clear that you don't need to group a single value - took me a second to work out what the length check meant ><

File: app/controllers/concerns/grouping_helper.rb [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    grouped_object.grouped_by = key
+    grouped_object.nodes = data_hash[key]
+
+    start_date = nil
+    end_date = nil
+    current_node = nil
+
+    grouped_object.nodes.each do |node|
+      # Find node that represents current role
+      # TODO: Might need to cater for multiple current nodes - eg. multiple current memberships and roles within a committee
```
Comment (mattrayner) [inline app/controllers/concerns/grouping_helper.rb]: is it worth implementing this now?

File: app/controllers/concerns/grouping_helper.rb [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    grouped_object.type = grouped_object.nodes.first.type
+    grouped_object.start_date = start_date
+    grouped_object.end_date = current_node.nil? ? end_date : nil
+
+    # Add sorted object to the array
+    sort_grouped_nodes_by_date(grouped_object, current_node)
+    grouped << grouped_object
+  end
+
+  def self.sort_grouped_nodes_by_date(grouped_object, current_node)
```
Comment (mattrayner) [inline app/controllers/concerns/grouping_helper.rb]: do you need to pass current_node in this instance? Could you not simplify by taking it straight from grouped_object?

File: app/controllers/concerns/history_helper.rb [hunk 14]
```diff
@@ -0,0 +1,64 @@
+module HistoryHelper
```
Comment (mattrayner) [inline app/controllers/concerns/history_helper.rb]: If we made this a class, we could remove the need to set `@history`, or call `.reset`?

What do you think?

File: app/controllers/concerns/history_helper.rb [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    # Set history if it has not already been set
+    history
+    data.each do |entry|
+      # Current
+      @history[:current] << entry if entry.try(:end_date).nil?
+
+      # Exit if there is no end date or if the end date is not a date
+      next if entry.try(:end_date).try(:year).try(:to_f).nil?
+
+      year = (((Time.now.year.to_f - entry.end_date.year.to_f) / year_seperator).ceil) * year_seperator
```
Comment (mattrayner) [inline app/controllers/concerns/history_helper.rb]: Might be worth giving a comment example with values to show the calculation?

**Pull request merged**